drivers/flash/jesd216.h

```diff
@@ -0,0 +1,416 @@
+	uint8_t wait_states;
+};
+
+/* Determine whether a particular operational mode is supported for
+ * read, and possibly what command may be used.
+ *
+ * @note For @p mode JESD216_MODE_111 this function will return zero
+ * to indicate that standard read (instruction 03h) is supported, but
+ * without providing information on how.  SFDP does not provide an
+ * indication of support for 1-1-1 Fast Read (0Bh).
+ *
+ * @param php pointer to the BFP header.
+ *
+ * @param bfp pointer to the BFP table.
+ *
+ * @param mode the desired protocol mode.
+ *
+ * @param res where to store instruction information.  Pass a null
+ * pointer to test for support without retrieving instruction
+ * information.
+ *
+ * @retval positive if instruction is supported and *res has been set.
+ *
+ * @retval 0 if instruction is supported but *res has not been set
+ * (e.g. no instruction needed, or instruction cannot be read from
+ * BFP).
+ *
+ * @retval -ENOTSUP if instruction is not supported.
+ */
+int jesd216_bfp_read_support(const struct jesd216_param_header *php,
+			     const struct jesd216_bfp *bfp,
+			     enum jesd216_mode_type mode,
+			     struct jesd216_instr *res);
+
+/* Description of a supported erase operation. */
+struct jesd216_erase_type {
+	/* The command opcode used for an erase operation. */
+	uint8_t cmd;
+
+	/* The value N when the erase operation erases a 2^N byte
+	 * region.
+	 */
+	uint8_t exp;
+};
+
+/* The number of erase types defined in a JESD216 Basic Flash
+ * Parameter table.
+ */
+#define JESD216_NUM_ERASE_TYPES 4
+
+/* Extract a supported erase size and command from BFP DW8 or DW9.
+ *
+ * @param bfp pointer to the parameter table.
+ *
+ * @param idx the erase type index, from 1 through 4.  Only index 1 is
+ * guaranteed to be present.
+ *
+ * @param etp where to store the command and size used for the erase.
+ *
+ * @retval 0 if the erase type index provided usable information.
+ * @retval -EINVAL if the erase type index is undefined.
+ */
+int jesd216_bfp_erase(const struct jesd216_bfp *bfp,
+		      uint8_t idx,
+		      struct jesd216_erase_type *etp);
+
+/* Extract typical and maximum erase times from DW10.
+ *
+ * @param php pointer to the BFP header.
+ *
+ * @param bfp pointer to the BFP table.
+ *
+ * @param idx the erase type index, from 1 through 4.
+ *
+ * @param typ_ms where to store the typical erase time (in
+ * milliseconds) for the specified erase type.
+ *
+ * @retval -ENOTSUP if the erase type index is undefined.
+ * @retval positive is a multiplier that converts typical erase times
+ * to maximum erase times.
+ */
+int jesd216_bfp_erase_type_times(const struct jesd216_param_header *php,
+				 const struct jesd216_bfp *bfp,
+				 uint8_t idx,
+				 uint32_t *typ_ms);
+
+/* Get the page size from the Basic Flash Parameters.
+ *
+ * @param php pointer to the BFP header
+ *
+ * @param bfp pointer to the BFP table
+ *
+ * @return the page size in bytes from the parameters if supported,
+ * otherwise 256.
```

Comment:
Nit pick regarding comments and documentation: sometimes you end comments for function parameters and structure members with dot, sometimes you do not (this vs `jesd216_bfp_erase_type_times`).

Reply:
For single-sentence descriptions I often don't add a period because it looks better when the fields are put into a table.  But no, it's not entirely consistent; I fixed up a few of them.

Reply:
OK, thanks.